NS-CACHE: WIP 2 - More caching scenarios

paulxuzhang opened the pull request:
### Explain the changes
Include PR #5977 changes done by @guymguym for caching. Additional changes is,
1. Add initial caching parameter to bucket API.
2. Add new property cache_valid_time to object for calculating TTL.
3. In read and head operations, cache valid time is updated if TTL expires and etag in hub and cache are same.
4. In read operation, object is read from hub and updated in cache if TTL expires and etag in hub and cache are different.
5. In read and head operations, 404 is returned and object is deleted from cache if it has been deleted from hub and TTL expires.
6. In upload operation, defer the finalizer of the cache stream until the hub upload completes. Clean up piped streams when error in read stream.
7. Update frontend to support creating namespace bucket with TTL. See 
![Screen Shot 2020-04-23 at 11 46 57 AM](https://user-images.githubusercontent.com/20991587/80807331-8913b180-8b82-11ea-9133-9ff95ddd78e0.png)
8. Added initial functional test file for namespace caching, namespace_cache_test.js.

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

guymguym commented:
@paulxuzhang Great! Will review.
So I should get my PR sorted out and ready to merge to master - for that I need to remove the hardcoded default to use namespace cache for every namespace bucket and use a config variable which will be only for dev right now.

paulxuzhang commented:
> @paulxuzhang Great! Will review.
> So I should get my PR sorted out and ready to merge to master - for that I need to remove the hardcoded default to use namespace cache for every namespace bucket and use a config variable which will be only for dev right now.

paulxuzhang commented:
@guymguym Regarding functional tests, I did not implement all test cases in the new file namespace_cache_test.js, I will create the rest of the tests later if the initial changes looks good.

nb-ohad commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.html:
```diff
@@ -86,6 +86,32 @@ <h2 class="heading3 push-prev">Write Policy</h2>
+                    <svg-icon class="icon-small valign-bottom"
+                        params="name: 'notif-info'"
+                    ></svg-icon>
+                    The cache TTL cannot be changed after set
```
Something in the phrasing here seems a little off. May I suggest:
```
This cache TTL value cannot be changed after the creation of the bucket.
```

paulxuzhang replied:
Ok

nb-ohad commented on frontend/src/app/epics/create-namespace-bucket.js:
```diff
@@ -10,12 +10,15 @@ export default function(action$, { api }) {
-                    write_resource: writeTo
+                    write_resource: writeTo,
+                    caching: {
+                        ttl: parseInt(cacheTTL)
```
`ParseInt` is not needed in this case because the field value should already be an integer.

paulxuzhang replied:
Cool.

nb-ohad commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.js:
```diff
@@ -206,6 +217,13 @@ class CreateNamespaceBucketModalViewModel extends ConnectableViewModel {
+            if (cacheTTL < 0) {
+                errors.cacheTTL = 'Cache TTL must be non-negative integer';
+
+            } else if (cacheTTL > 172800) {
```
The max value here is different then the max value as defined in the bucket API schema: 1296000. I am not sure if this is intentional or not.

paulxuzhang replied:
I will figure out and use the same limit.

nb-ohad commented on src/endpoint/endpoint.js:
```diff
@@ -70,12 +70,13 @@ function process_env(env) {
+        md: env.MD_ADDR || `${default_base_addr}:${default_base_port + 1}`,
+        bg: env.BG_ADDR || `${default_base_addr}:${default_base_port + 2}`,
+        hosted_agents: env.HOSTED_AGENTS_ADDR || `${default_base_addr}:${default_base_port + 3}`,
+        master: env.MGMT_ADDR || `${default_base_addr}:${default_base_port}`
```
`addr_utils.get_default_ports(Number(env.SSL_PORT) || 8443)` can provide all the ports values for all the services. And is a single place to change in case we decide to change the logic of the port selection.

guymguym replied:
This is from my branch. Will update.

paulxuzhang replied:
pl

paulxuzhang replied:
@guymguym When will you be able to merge your namespace caching PR?

paulxuzhang commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.js:
```diff
@@ -111,6 +114,7 @@ class CreateNamespaceBucketModalViewModel extends ConnectableViewModel {
         const bucketName = getFieldValue(form, 'bucketName');
         const readPolicy = getFieldValue(form, 'readPolicy');
         const writePolicy = getFieldValue(form, 'writePolicy');
+        const cacheTTL = getFormValues(form).cacheTTL;
```
Ok. I will probably go with
```
const { bucketName, readPolicy, writePolicy, cacheTTL } = getFormValues(form) || {};
```

paulxuzhang commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.js:
```diff
@@ -172,6 +177,12 @@ class CreateNamespaceBucketModalViewModel extends ConnectableViewModel {
         this
     );
 
+    onCacheTTL = throttle(
```
Ok

paulxuzhang commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.js:
```diff
@@ -206,6 +217,13 @@ class CreateNamespaceBucketModalViewModel extends ConnectableViewModel {
             }
+        } else if (step === 2) {
+            if (cacheTTL < 0) {
+                errors.cacheTTL = 'Cache TTL must be non-negative integer';
```
Ok

paulxuzhang commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.html:
```diff
@@ -86,6 +86,32 @@ <h2 class="heading3 push-prev">Write Policy</h2>
+                <editor class="greedy" params="label: 'Cache TTL'">
+                    <input type="number" min="0" max="172800"
+                        class="push-next-half"
+                        ko.textInput="ko.pc($form.cacheTTL, onCacheTTL)"
```
Ok

nb-ohad commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.html:
```diff
@@ -86,6 +86,32 @@ <h2 class="heading3 push-prev">Write Policy</h2>
+                    Caching configuration controls how data is cached in local storage
+                </p>
+                <editor class="greedy" params="label: 'Cache TTL'">
+                    <input type="number" min="0" max="172800"
```
The max value of the input field does not match the new value in the `onValidate` method and `bucket_api` ( which is 3600 after the last changes).

You can add a property on the component class named `maxCacheTTL = 3600` (in the property declaration section`  and use it both in `onValidate` and here, by replacing `mak=172800` with `ko.attr.max=maxCacheTTL` os this value will be controlled from a single place.  

Please note that this value does not need to be `ko.observable` just a plain value (it is supposed to be constant)

paulxuzhang replied:
Ok.

nb-ohad reviewed (approved):
Reviewed mostly FE changes

paulxuzhang commented:
@guymguym Rebased. Could you review?

paulxuzhang commented:
@guymguym Changed the TTL in bucket API to be in milliseconds.

paulxuzhang commented:
@guymguym @jeniawhite Any issue?

guymguym commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.html:
```diff
@@ -86,6 +86,34 @@ <h2 class="heading3 push-prev">Write Policy</h2>
+            <div class="column greedy pad">
+                <h2 class="heading3">Caching Configuration</h2>
+                <p class="push-next">
+                    Caching configuration controls how data is cached in local storage
```
So perhaps we need to have a ui-toggle that opens this section on explicitly in order to enable caching?

paulxuzhang replied:
I will make it optional. I am not UI developer. It would have to be polished.

guymguym commented on src/server/object_services/object_server.js:
```diff
@@ -399,6 +399,7 @@ async function complete_object_upload(req) {
     }
 
     set_updates.create_time = new Date();
+    set_updates.cache_valid_time = new Date();
```
Why are we adding this property unconditionally?
I expect it to be added only for caching buckets:
```suggestion
    if (req.bucket.namespace.caching) {
        set_updates.cache_valid_time = new Date();
    }
```

paulxuzhang replied:
Ok. Will make it conditional.

guymguym commented on src/server/system_services/bucket_server.js:
```diff
@@ -153,13 +153,20 @@ async function create_bucket(req) {
         if (req.rpc_params.namespace.write_resource && !write_resource) {
             throw new RpcError('INVALID_WRITE_RESOURCES');
         }
+        let caching = {
```
Default should be undefined, and only set the caching config object when this bucket has enabled caching:
```suggestion
        let caching;
```

paulxuzhang replied:
ok

guymguym commented on src/server/system_services/bucket_server.js:
```diff
@@ -1400,7 +1410,8 @@ function get_bucket_info({
                 bucket.namespace.write_resource).name,
-            read_resources: _.map(bucket.namespace.read_resources, rs => pool_server.get_namespace_resource_info(rs).name)
+            read_resources: _.map(bucket.namespace.read_resources, rs => pool_server.get_namespace_resource_info(rs).name),
+            caching: bucket.namespace.caching ? bucket.namespace.caching : { ttl: 0 }
```
Return undefined if undefined:
```suggestion
            caching: bucket.namespace.caching,
```

paulxuzhang replied:
ok

guymguym commented on src/server/system_services/bucket_server.js:
```diff
@@ -468,6 +475,9 @@ async function read_bucket_sdk_info(req) {
             read_resources: _.map(bucket.namespace.read_resources, rs => pool_server.get_namespace_resource_extended_info(rs))
         };
+        if (bucket.namespace.caching) {
+            reply.namespace.caching = bucket.namespace.caching;
```
No need to separate the new property from the others:
```suggestion
        reply.namespace = {
            write_resource: pool_server.get_namespace_resource_extended_info(
                system.namespace_resources_by_name[bucket.namespace.write_resource.name]
            ),
            read_resources: _.map(bucket.namespace.read_resources, rs => pool_server.get_namespace_resource_extended_info(rs))
            caching: bucket.namespace.caching,
        };
```

paulxuzhang replied:
ok

guymguym reviewed (changes requested):
@paulxuzhang Hi, see comments above.
I still need to go over the flows in namespace_cache.

paulxuzhang commented on src/server/object_services/schemas/object_md_schema.js:
```diff
@@ -76,6 +76,7 @@ module.exports = {
         upload_size: { type: 'integer' },
         upload_started: { objectid: true },
         create_time: { date: true },
+        cache_valid_time: { date: true },
```
ok

paulxuzhang commented on src/api/object_api.js:
```diff
@@ -254,6 +254,9 @@ module.exports = {
                     create_time: {
                         idate: true
                     },
+                    cache_valid_time: {
```
It shouldn't be returned for multipart complete. Will remove.

paulxuzhang commented on src/api/bucket_api.js:
```diff
@@ -1129,6 +1132,7 @@ module.exports = {
             required: [
                 'read_resources',
                 'write_resource',
+                'caching'
```
ok

paulxuzhang commented on src/api/common_api.js:
```diff
@@ -748,6 +748,25 @@ module.exports = {
                 'NO_RESOURCES'
             ]
         },
+
+        bucket_cache_ttl: {
+            type: 'integer',
+            // In seconds, 0 means caching is disabled
+            minimum: 0,
+            // 1 hour in milliseconds
+            maximum: 3600000
+        },
+
+        bucket_cache_config: {
+            type: 'object',
+            required: [ 'ttl' ],
+            properties: {
+                ttl: {
+                    $ref: '#/definitions/bucket_cache_ttl'
+                }
+            }
+        },
+
```
ok

paulxuzhang commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -15,14 +15,21 @@ async function get_object(req, res) {
-    const object_md = await req.object_sdk.read_object_md({
+    let get_from_cache = false;
+    if (req.query.get_from_cache === 'true') {
+        get_from_cache = true;
+    }
+    let md_params = {
         bucket: req.params.bucket,
         key: req.params.key,
         version_id: req.query.versionId,
         md_conditions: http_utils.get_md_conditions(req),
         encryption
-    });
-
+    };
+    if (get_from_cache) {
+        md_params.get_from_cache = true;
+    }
+    const object_md = await req.object_sdk.read_object_md(md_params);
```
ok

paulxuzhang replied:
@guymguym I have to take back my previous answer.  If we made the changes you proposed, then we would have to change other namespace files (e,g, namespace_nb.js) for removing get_from_cache from params passed to API read_object_md and read_object_stream, since get_from_cache is not in the API schema.
Even after replacing ```get_from_cache: req.query.get_from_cache === 'true'``` with ```get_from_cache: req.query.get_from_cache === 'true' ? true : undefined```, it will not work either without changes in other places, since rpc schema validation takes undefined field as present.

guymguym replied:
OK so our schemas are strict and we inject `additionalProperties: false` by default to all of our schemas.
See strictify in schema_utils:
https://github.com/noobaa/noobaa-core/blob/9fe155f0191742fc0ee4d4eaad1da79fec47622a/src/util/schema_utils.js#L137

This causes ajv to fail if the key exists even if it is undefined.
Opened an issue on AJV - https://github.com/ajv-validator/ajv/issues/1240
Resolving this comment for now until we have a better way.

paulxuzhang commented on src/endpoint/s3/ops/s3_head_object.js:
```diff
@@ -10,13 +10,18 @@ const http_utils = require('../../../util/http_utils');
+    if (req.query.get_from_cache === 'true') {
+        params.get_from_cache = true;
+    }
+    const object_md = await req.object_sdk.read_object_md(params);
```
ok

paulxuzhang replied:
See my response in the issue on s3_get_object.

guymguym replied:
I'm resolving this one and keep the discussion only on s3_get_object so whatever is done there should be applied here.

paulxuzhang commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.html:
```diff
@@ -86,6 +86,34 @@ <h2 class="heading3 push-prev">Write Policy</h2>
+                    <svg-icon class="icon-small valign-bottom"
+                        params="name: 'notif-info'"
+                    ></svg-icon>
+                    This cache TTL value cannot be changed after the creation of the bucket.
```
Update scenario is not included in this PR. Haven't figured out how to implement update in the frontend UI. Should we support enabling/disabling caching besides supporting ttl update?

Suggest support update scenarios in separate PRs.
- changes in API and bucket_server
- Frontend UI changes

guymguym replied:
Yes I agree, let's keep it off this PR.
Good questions. I would think that updating the TTL should be supported, but we don't want to allow switching caching on/off for now. Certainly not from the UI. Maybe we can leave it open to the API so that we can experiment with.

guymguym replied:
Resolving.

paulxuzhang commented on frontend/src/app/epics/create-namespace-bucket.js:
```diff
@@ -10,12 +10,16 @@ export default function(action$, { api }) {
                     read_resources: readFrom,
-                    write_resource: writeTo
+                    write_resource: writeTo,
+                    caching: {
```
I will give it try to make it optional.

guymguym commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -15,14 +15,21 @@ async function get_object(req, res) {
-    });
-
+    };
+    if (get_from_cache) {
+        md_params.get_from_cache = true;
+    }
```
```suggestion
    if (req.query.get_from_cache !== undefined) {
        params.get_from_cache = true;
    }
```

You don't need to go through another local variable `get_from_cache`, and you don't need to require `req.query.get_from_cache` to have a specific literal string value of `"true"` because it is easy to forget and use the simple form of `https://endpoint?get_from_cache` which actually gets parsed to an empty string:

```
$ node -p 'querystring.parse("get_from_cache&x=true&x=true")'
[Object: null prototype] { get_from_cache: '', x: [ 'true', 'true' ] }
```

So checking that `query.key` is not undefined means all the url query options work, even accidental var repetition that will result in an array.

paulxuzhang replied:
ok

guymguym commented on src/endpoint/s3/ops/s3_get_object.js:
```diff
@@ -15,14 +15,21 @@ async function get_object(req, res) {
+    if (req.query.get_from_cache === 'true') {
+        get_from_cache = true;
+    }
+    let md_params = {
```
```suggestion
    const params = {
```
You should still use `const` and not `let` because const only applies to the reference in javascript and not to the object itself. For objects to be immutable you use `Object.freeze(obj)` or even `js_utils.deep_freeze(obj)`.

Also the name `params` would be more compatible with the name inside the call.

paulxuzhang replied:
ok

guymguym commented on src/sdk/namespace_cache.js:
```diff
@@ -136,12 +201,30 @@ class NamespaceCache {
-            const [hub_res, cache_res] = await Promise.all([hub_promise, cache_promise]);
-            assert.strictEqual(hub_res.etag, cache_res.etag);
-            return hub_res;
+            const hub_upload = await Promise.all([ hub_promise, cache_promise ])
+            .then(([hub_res, cache_res]) => {
+                assert.strictEqual(hub_res.etag, cache_res.etag);
+                return hub_res;
+            })
+            .catch(err => {
+                dbg.log0("NamespaceCache.upload_object: error in upload", {params, error: err});
+                // If error is from cache, we should probably ignore and let hub upload continue. Change Promise.all to Promise.allSettled?
+                throw err;
+            });
```
This error handling wouldn't really work - you need to use try-catch around await.
Also we need to return just a single result and not the array of two results.
I agree with the comment you wrote and we should change it to allSettled.

See suggested code -

```suggestion
            const [hub_res, cache_res] = await Promise.allSettled([ hub_promise, cache_promise ]);
            const hub_ok = hub_res.status === 'fulfilled';
            const cache_ok = cache_res.status === 'fulfilled';

            if (!hub_ok) {
                dbg.log0("NamespaceCache.upload_object: error in upload", params, hub_res, cache_res);

                // handling the case where cache succeeded and cleanup.
                // We can also just mark the cache object for re-validation
                // to make sure any read will have to re-validate it,
                // but writes (retries of the upload most likely) will be already in the cache
                // and detected by dedup so we don't need to do anything.
                if (cache_ok) {
                    try {
                        await this. _delete_object_from_cache(params, object_sdk);
                    } catch (err) {
                        dbg.error('Failed to delete object from cache after hub failure', err.stack);
                    }
                }

                // fail back to client with the hub reason
                throw hub_res.reason;
            }
            
            if (cache_ok) {
                assert.strictEqual(hub_res.value.etag, cache_res.value.etag);
            } else {
                // on error from cache, we ignore and let hub upload continue
                dbg.log0("NamespaceCache.upload_object: error in cache upload", params, hub_res, cache_res);
            }

            return hub_res.value;
```

paulxuzhang replied:
ok

paulxuzhang commented on src/sdk/namespace_cache.js:
```diff
@@ -4,19 +4,35 @@
         return this.namespace_hub;
     }
 
+    _delete_object_from_cache(params, object_sdk) {
```
ok

paulxuzhang commented on src/sdk/namespace_cache.js:
```diff
@@ -4,19 +4,35 @@
     }
 
+    _delete_object_from_cache(params, object_sdk) {
+        process.nextTick(() => {
```
Since cache delete does not need to block s3 procedure, client can get quicker response from the endpoint without waiting for cache delete to be finished. Because no await is used, by using nextTick, it allows the cache delete to be handled as soon as possible by node ( after the current operation in node is completed), so cache entry will be cleaned up before client's next s3 operation.

guymguym replied:
First nextTick should probably be replaced with setImmediate because nextTick shouldn't be used unless no other option because it can starve IO events.
But in this case to submit an operation in the background, we usually write the async function as usual, so that if some callers want to await it they can do so, and for callers that do not want to wait we just call without awaiting, and the effect is the same. You could also do this on the call side to make it more visible:
```
setImmediate(() => this. _delete_object_from_cache(...));
```

paulxuzhang replied:
Ok. Will use try catch.

paulxuzhang commented on src/sdk/namespace_cache.js:
```diff
@@ -4,19 +4,35 @@
+        process.nextTick(() => {
+            const delete_params = _.pick(params, 'bucket', 'key');
+            dbg.log0("NamespaceCache._delete_object_from_cache", delete_params);
+            this.namespace_nb.delete_object(delete_params, object_sdk)
+            .then(() => {
+                dbg.log0('NamespaceCache: deleted object from cache', delete_params);
+            })
+            .catch(err_delete => {
+                dbg.warn('NamespaceCache: error in deleting object from cache', params, err_delete);
+            });
```
It depends on whether we would like to have s3 operation blocked by cache cleanup. See my response in https://github.com/noobaa/noobaa-core/pull/6004#discussion_r447018283

guymguym replied:
Actually it doesn't really depend - you can still make the callback function invoked by nextTick to be async, clearly nextTick will not await for it, but it will allow your function to be written with await and try-catch inside which makes it easier to read

paulxuzhang replied:
Will change it to use try and catch.
Will NOT use await to call _delete_object_from_cache() in relevant places.

paulxuzhang commented on src/server/system_services/bucket_server.js:
```diff
@@ -154,12 +154,19 @@ async function create_bucket(req) {
             throw new RpcError('INVALID_WRITE_RESOURCES');
         }
 
+        let caching;
+        if (req.rpc_params.namespace.caching) {
+            caching = req.rpc_params.namespace.caching;
+            caching.ttl_ms = _.defaultTo(caching.ttl_ms, config.NAMESPACE_CACHING.DEFAULT_CACHE_TTL_MS);
+        }
+
```
Cool. Good to know. I will make changes.

paulxuzhang commented on frontend/src/app/epics/create-namespace-bucket.js:
```diff
@@ -10,12 +10,19 @@ export default function(action$, { api }) {
+            const { name, readFrom, writeTo, cacheTTL } = action.payload;
 
+            let caching;
+            if (cacheTTL != undefined) {
```
ok

paulxuzhang commented on frontend/src/app/components/modals/create-namespace-bucket-modal/create-namespace-bucket-modal.js:
```diff
@@ -237,10 +245,10 @@ class CreateNamespaceBucketModalViewModel extends ConnectableViewModel {
         this.dispatch(
             closeModal(),
-            createNamespaceBucket(bucketName, readPolicy, writePolicy)
+            createNamespaceBucket(bucketName, readPolicy, writePolicy, this.enableCaching() ? cacheTTL : undefined)
```
ok

guymguym commented on src/test/utils/s3ops.js:
```diff
@@ -673,15 +674,29 @@ class S3OPS {
+            const obj = await this.s3.getObject(params)
+            .on('build', req => {
+                if (query_params) {
+                    for (const [k, v] of Object.entries(query_params)) {
```
Use querystring module:
```
const querystring = require('querystring');
querystring.stringify(query_params);
```

paulxuzhang replied:
ok

jeniawhite commented on src/sdk/object_sdk.js:
```diff
@@ -163,15 +163,14 @@ class ObjectSDK {
                         ns: new NamespaceCache({
                             namespace_hub: this._setup_single_namespace(_.extend({}, bucket.namespace.write_resource)),
                             namespace_nb: this.namespace_nb,
                             active_triggers: bucket.namespace.write_resource,
```
@guymguym
You should pass the active_triggers that you've received from read_bucket_sdk_info.
We have support for triggers in namespace_nb and namespace_merge (supports s3, blob etc...).
Since namespace_cache is essentially like namespace_merge (not like namespace_s3/blob/etc...), you will need to call the `object_sdk.should_run_triggers` and `object_sdk.dispatch_triggers` methods from the namespace_cache in the relevant paths.

paulxuzhang replied:
I will make the changes.

paulxuzhang replied:
@guymguym @jeniawhite I made the changes by following namespace_merge. Could you review?

guymguym replied:
looks ok to me, but mostly I wonder why we can't move the triggers code up one level to object_sdk around where it calls the namespace functions, and then have it in one place for all namespaces.

paulxuzhang replied:
What are the namespaces supporting active_triggers? Are they nb, merge and cache so far? By glancing through the code, it seems that they have different logic in active_triggers, for example,
- nb has trigger in complete_object_upload, but merge does not have trigger in complete_object_upload.
- merge has logic in handling multiple read sources.

paulxuzhang replied:
Found that there is bug in my original changes related to active triggers in delete_multiple_objects(). For now, my opinion is not to move the logic to object_sdk since the code is not much common.

guymguym replied:
OK so for now we can use the same code copied from namespace_merge, but going forward we might be able to have a default code applied by object_sdk and the namespace will override if needed for optimizing specific cases where we already have all the information locally and do not need to fetch it. I am resolving this comment.

guymguym commented on src/test/utils/s3ops.js:
```diff
@@ -673,15 +675,22 @@ class S3OPS {
-            const obj = await this.s3.getObject(params).promise();
+            const obj = await this.s3.getObject(params)
+            .on('build', req => {
+                if (query_params) {
+                    const query_string = querystring.stringify(query_params);
+                    console.log(`Setting query parameter ${query_string}`);
+                    req.httpRequest.path += `?${query_string}`;
+                }
```
I think you still need to use the same condition as you had before in case the path already contains a search string, something like:
```suggestion
                if (!query_params) return;
                const sep = req.httpRequest.search() ? '&' : '?';
                const query_string = querystring.stringify(query_params);
                const req_path = `${req.httpRequest.path}${sep}${query_string}`;
                console.log(`Added query params ${query_params} to path ${req_path}`);
                req.httpRequest.path = req_path;
```

paulxuzhang replied:
My bad. Should have checked whether request has existing query parameter like before.

liranmauda commented on src/test/pipeline/namespace_cache_test.js:
```diff
@@ -0,0 +1,493 @@
+            }
+        }
+    }
+    await _delete_namesapace_bucket(bucket);
```
Typo:
```suggestion
    await _delete_namespace_bucket(bucket);
```

paulxuzhang replied:
ok

liranmauda commented on src/test/pipeline/namespace_cache_test.js:
```diff
@@ -0,0 +1,493 @@
+    }
+}
+
+async function _delete_namesapace_bucket(bucket) {
```
Typo:
```suggestion
async function _delete_namespace_bucket(bucket) {
```

paulxuzhang replied:
Good catch.

liranmauda commented on src/test/pipeline/namespace_cache_test.js:
```diff
@@ -0,0 +1,493 @@
+    ip: 's3.us-south.cloud-object-storage.appdomain.cloud',
+    access_key: COSDefaultConnection.identity,
+    secret_key: COSDefaultConnection.secret,
+    system_verify_name: 'AWS',
```
Should this be COS? 
```suggestion
    system_verify_name: 'COS',
```

paulxuzhang replied:
ok

liranmauda commented on src/test/pipeline/namespace_cache_test.js:
```diff
@@ -0,0 +1,493 @@
+
+async function compereMD5betweenCloudAndNooBaa(type, cloud_bucket, noobaa_bucket, force_cache_read, file_name) {
+    console.log(`Compering NooBaa cache bucket to ${type} for ${file_name}`);
+    let cloudProperties;
+    cloudProperties = await getPropertiesFromS3Bucket(namespace_mapping[type].s3ops, cloud_bucket, file_name, false);
```
```suggestion
    const cloudProperties = await getPropertiesFromS3Bucket(namespace_mapping[type].s3ops, cloud_bucket, file_name, false);
```

paulxuzhang replied:
ok

liranmauda commented on src/test/pipeline/namespace_cache_test.js:
```diff
@@ -0,0 +1,493 @@
+}
+
+async function set_rpc_and_create_auth_token() {
+    let auth_params = {
```
```suggestion
    const auth_params = {
```

paulxuzhang replied:
ok

guymguym commented on src/test/utils/s3ops.js:
```diff
@@ -37,6 +37,7 @@ class S3OPS {
             throw new Error('You cannot do this: SigV4 requires https to disable body signing and send streams...');
         }
         const rest_endpoint = use_https ? `https://${ip}:${ssl_port}` : `http://${ip}:${port}`;
+        console.log("rest_endpoint:", rest_endpoint);
```
@liranmauda no need for space there since console adds a space between arguments.

guymguym commented on src/sdk/namespace_cache.js:
```diff
@@ -41,43 +54,104 @@ class NamespaceCache {
+            object_info_hub = await this.namespace_hub.read_object_md(params, object_sdk);
+            if (object_info_hub.etag === cache_etag) {
+                dbg.log0('NamespaceCache.read_object_md: same etags: updating cache valid time', object_info_hub);
+                process.nextTick(() => {
```
Same comments as before - use an internal async method on the namespace class such as `async _update_cache_last_valid_time()` which will do its internal try-catch-await and then call it from here with `setImmediate(() => this.__update_cache_last_valid_time(...))`

paulxuzhang replied:
ok

guymguym commented on src/sdk/namespace_cache.js:
```diff
@@ -41,43 +54,104 @@ class NamespaceCache {
-        return this.namespace_hub.read_object_md(params, object_sdk);
+        let object_info_hub = null;
+        try {
+            object_info_hub = await this.namespace_hub.read_object_md(params, object_sdk);
```
Since you use this only to return result in case of success, doesn't seem like you need to have this as a mutable `let` variable and can use a scoped `const` variable.
```suggestion
            const object_info_hub = await this.namespace_hub.read_object_md(params, object_sdk);
```

paulxuzhang replied:
ok

guymguym reviewed (approved):
He @paulxuzhang 
OK the PR is looking good!
Please resolve the last comments and rebase so we can merge later today.
Godspeed!

paulxuzhang commented on src/endpoint/s3/ops/s3_head_object.js:
```diff
@@ -10,13 +10,18 @@ const http_utils = require('../../../util/http_utils');
 async function head_object(req, res) {
     const encryption = s3_utils.parse_encryption(req);
-    const object_md = await req.object_sdk.read_object_md({
+    let params = {
```
Already changed.

paulxuzhang commented on src/test/utils/bucket_functions.js:
```diff
@@ -151,18 +151,20 @@ class BucketFunctions {
+    async createNamespaceBucket(name, namespace, caching) {
+        console.log(`Creating namespace bucket ${name} with namespace ${namespace}`);
         try {
+            let namespaceConfig = {
```
ok

paulxuzhang committed:
Namespace caching: support cache TTL and related scenario

- Add new property cache_valid_time to object for calculating TTL
- In read and head operations, cache valid time is updated if TTL expires and etag in hub and cache are same
- In read operation, object is read from hub and updated in cache if TTL expires and etag in hub and cache are different.
- In read and head operations, 404 is returned and object is deleted from cache if it has been deleted from hub and TTL expires.
- Update frontend to support creating namespace bucket with TTL

paulxuzhang commented:
@guymguym Resolved the last comments and rebased. But the unit tests failed due to hang. Trying to figure it out.

paulxuzhang commented:
@guymguym Unit tests pass now after restarted.
